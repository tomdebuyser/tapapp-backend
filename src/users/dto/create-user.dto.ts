import { IsEmail, IsString, IsOptional } from 'class-validator';

export class CreateUserRequest {
    @IsEmail()
    readonly email: string;

    @IsString()
    @IsOptional()
    readonly firstName?: string;

    @IsString()
    @IsOptional()
    readonly lastName?: string;
}
