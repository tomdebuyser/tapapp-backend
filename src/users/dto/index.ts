export * from './create-user.dto';
export * from './get-users.dto';
export * from './resend-register-mail.dto';
