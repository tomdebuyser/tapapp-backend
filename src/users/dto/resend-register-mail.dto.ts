import { IsUUID } from 'class-validator';
import { ApiProperty } from '@nestjs/swagger';

export class ResendRegisterMailRequestParams {
    @ApiProperty({ required: true })
    @IsUUID('4')
    readonly userId: string;
}
