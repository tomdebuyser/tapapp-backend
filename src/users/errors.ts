import { BadRequestException } from '@nestjs/common';

export class EmailAlreadyInUse extends BadRequestException {
    constructor() {
        super('A user with this email already exists', 'EMAIL_ALREADY_IN_USE');
    }
}

export class RoleNotFound extends BadRequestException {
    constructor() {
        super('One of the roles passed was not found', 'ROLE_NOT_FOUND');
    }
}
