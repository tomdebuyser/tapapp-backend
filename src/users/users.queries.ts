import { Injectable } from '@nestjs/common';
import { mergeDeepLeft } from 'ramda';

import {
    GetUsersRequestQuery,
    GetUsersResponse,
    UsersSortColumns,
} from './dto';
import { UserRepository } from '../database';
import { SortDirection } from '../_shared/constants';

@Injectable()
export class UsersQueries {
    constructor(private readonly userRepository: UserRepository) {}

    async getUsers(
        requestQuery: GetUsersRequestQuery,
    ): Promise<GetUsersResponse> {
        const defaultQuery: GetUsersRequestQuery = {
            skip: 0,
            take: 20,
            sortBy: UsersSortColumns.UpdatedAt,
            sortDirection: SortDirection.Descending,
            search: '',
        };
        const query = mergeDeepLeft(requestQuery, defaultQuery);

        const [users, totalCount] = await this.userRepository
            .createQueryBuilder('user')
            .select([
                'user.id',
                'user.createdAt',
                'user.updatedAt',
                'user.createdBy',
                'user.updatedBy',
                'user.email',
                'user.state',
                'user.firstName',
                'user.lastName',
                'role.id',
                'role.name',
            ])
            .innerJoin('user.roles', 'role')
            .orderBy(`user.${query.sortBy}`, query.sortDirection)
            .take(query.take)
            .skip(query.skip)
            .where('user.email ILIKE :search', {
                search: `%${query.search}%`,
            })
            .getManyAndCount();

        return {
            meta: {
                count: users.length,
                totalCount,
                skip: query.skip,
            },
            data: users,
        };
    }
}
