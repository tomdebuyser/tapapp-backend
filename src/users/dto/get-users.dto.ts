import { IsOptional, IsEnum } from 'class-validator';
import { Type } from 'class-transformer';

import { PagingQuery, PagingMeta } from '../../_shared/dto';
import { UserState } from '../../_shared/constants';

export enum UsersSortColumns {
    Email = 'email',
    State = 'state',
    CreatedAt = 'createdAt',
    UpdatedAt = 'updatedAt',
}

export class GetUsersRequestQuery extends PagingQuery {
    @IsOptional()
    @IsEnum(UsersSortColumns)
    sortBy?: UsersSortColumns;
}

export class GetUsersResponse {
    meta: PagingMeta;
    data: GetUsersResponseData[];
}

class GetUsersResponseData {
    id: string;
    @Type(() => String)
    createdAt: Date;
    @Type(() => String)
    updatedAt: Date;
    email: string;
    state: UserState;
    firstName?: string;
    lastName?: string;
}
