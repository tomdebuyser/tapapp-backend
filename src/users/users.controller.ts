import {
    Controller,
    Post,
    Body,
    Get,
    Query,
    UseGuards,
    Param,
    HttpCode,
    HttpStatus,
    Patch,
} from '@nestjs/common';
import { ApiTags } from '@nestjs/swagger';

import {
    CreateUserRequest,
    GetUsersResponse,
    GetUsersRequestQuery,
    UpdateUserRequest,
    UserIdParam,
    UserResponse,
} from './dto';
import { UsersService } from './users.service';
import { UsersQueries } from './users.queries';
import { AuthenticatedGuard } from '../_shared/guards';
import { Origin, UserSession } from '../_shared/decorators';
import { IUserSession } from '../_shared/constants';

@UseGuards(AuthenticatedGuard)
@ApiTags('users')
@Controller('users')
export class UsersController {
    constructor(
        private readonly usersService: UsersService,
        private readonly usersQueries: UsersQueries,
    ) {}

    @Get()
    getUsers(@Query() query: GetUsersRequestQuery): Promise<GetUsersResponse> {
        return this.usersQueries.getUsers(query);
    }

    @Post()
    async createUser(
        @Body() body: CreateUserRequest,
        @UserSession() session: IUserSession,
        @Origin() origin: string,
    ): Promise<UserResponse> {
        const userId = await this.usersService.createUser(
            body,
            session,
            origin,
        );
        return this.usersQueries.getUser(userId);
    }

    @Patch(':userId')
    async updateUser(
        @Body() body: UpdateUserRequest,
        @Param() params: UserIdParam,
        @UserSession() session: IUserSession,
    ): Promise<UserResponse> {
        // Only the provided properties will be updated
        const userId = await this.usersService.updateUser(
            body,
            params.userId,
            session,
        );
        return this.usersQueries.getUser(userId);
    }

    @HttpCode(HttpStatus.NO_CONTENT)
    @Post(':userId/resend-register-mail')
    resendRegisterMail(
        @Param() params: UserIdParam,
        @UserSession() session: IUserSession,
        @Origin() origin: string,
    ): Promise<void> {
        return this.usersService.resendRegisterMail(
            params.userId,
            session,
            origin,
        );
    }
}
