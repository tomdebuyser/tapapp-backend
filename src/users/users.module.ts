import { Module } from '@nestjs/common';
import { JwtModule } from '@nestjs/jwt';

import { UsersController } from './users.controller';
import { UsersService } from './users.service';
import { DatabaseModule } from '../database';
import { Config } from '../config';

@Module({
    imports: [
        DatabaseModule,
        JwtModule.register({
            secret: Config.jwt.secret,
        }),
    ],
    controllers: [UsersController],
    providers: [UsersService],
})
export class UsersModule {}
