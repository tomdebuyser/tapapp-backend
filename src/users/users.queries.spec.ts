import { Test } from '@nestjs/testing';

import { UsersQueries } from './users.queries';
import { DatabaseModule, UserRepository } from '../database';
import { GetUsersRequestQuery, UsersSortColumns } from './dto';
import { SortDirection } from '../_shared/constants';

describe('UsersQueries', () => {
    let userRepository: UserRepository;
    let usersQueries: UsersQueries;

    beforeAll(async () => {
        const module = await Test.createTestingModule({
            imports: [DatabaseModule],
            providers: [UsersQueries],
        }).compile();

        usersQueries = module.get(UsersQueries);
        userRepository = module.get(UserRepository);
    });

    afterAll(() => {
        userRepository.manager.connection.close();
    });

    describe('getUsers', () => {
        const testQueries: GetUsersRequestQuery[] = [
            {},
            { skip: 2, take: 2 },
            { search: 'user1' },
            {
                sortBy: UsersSortColumns.CreatedAt,
                sortDirection: SortDirection.Ascending,
            },
        ];

        testQueries.forEach((query, index) => {
            it(`should return a paged list of users: #${index}`, async () => {
                const result = await usersQueries.getUsers(query);
                expect(result).toMatchSnapshot();
            });
        });

        it('should return an empty list if no users found', async () => {
            const result = await usersQueries.getUsers({
                search: 'nonsensicaljibberish',
            });

            expect(result).toMatchSnapshot();
        });
    });
});
