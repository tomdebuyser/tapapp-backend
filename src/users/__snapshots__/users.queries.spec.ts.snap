// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersQueries getUsers should return a paged list of users: #0 1`] = `
Object {
  "data": Array [
    User {
      "createdAt": 2020-01-01T10:21:59.730Z,
      "email": "user2@silvernext.com",
      "firstName": "Benno",
      "id": "45c2c554-b06a-4209-aba8-b097cb5f8566",
      "lastName": "Meysmans",
      "state": "ACTIVE",
      "updatedAt": 2020-01-05T08:21:59.730Z,
    },
    User {
      "createdAt": 2020-01-02T09:21:59.730Z,
      "email": "user5@silvernext.com",
      "firstName": null,
      "id": "af965f04-8c7d-4e20-b230-932b0ff3a33f",
      "lastName": null,
      "state": "REGISTERING",
      "updatedAt": 2020-01-02T09:21:59.730Z,
    },
    User {
      "createdAt": 2019-01-01T08:32:59.730Z,
      "email": "user4@silvernext.com",
      "firstName": null,
      "id": "c356fce8-9f95-4c69-9e71-8f4d74f49b8f",
      "lastName": null,
      "state": "ACTIVE",
      "updatedAt": 2020-01-02T08:59:59.730Z,
    },
    User {
      "createdAt": 2020-01-01T08:21:59.730Z,
      "email": "development@icapps.com",
      "firstName": null,
      "id": "c4cb4582-1e97-4e3e-9d49-c744c8c1a229",
      "lastName": null,
      "state": "ACTIVE",
      "updatedAt": 2020-01-01T08:21:59.730Z,
    },
    User {
      "createdAt": 2020-01-01T08:21:59.730Z,
      "email": "user1@silvernext.com",
      "firstName": "Tom",
      "id": "49affd08-f461-4e80-899f-7612e971ff12",
      "lastName": "De Buyser",
      "state": "ACTIVE",
      "updatedAt": 2020-01-01T08:21:59.730Z,
    },
    User {
      "createdAt": 2020-01-01T08:21:59.730Z,
      "email": "user3@silvernext.com",
      "firstName": "Jens",
      "id": "db5e788c-2642-4c40-a63a-f13675048ddc",
      "lastName": "De Craecker",
      "state": "REGISTERING",
      "updatedAt": 2020-01-01T08:21:59.730Z,
    },
  ],
  "meta": Object {
    "count": 6,
    "totalCount": 6,
  },
}
`;

exports[`UsersQueries getUsers should return a paged list of users: #1 1`] = `
Object {
  "data": Array [
    User {
      "createdAt": 2019-01-01T08:32:59.730Z,
      "email": "user4@silvernext.com",
      "firstName": null,
      "id": "c356fce8-9f95-4c69-9e71-8f4d74f49b8f",
      "lastName": null,
      "state": "ACTIVE",
      "updatedAt": 2020-01-02T08:59:59.730Z,
    },
    User {
      "createdAt": 2020-01-01T08:21:59.730Z,
      "email": "development@icapps.com",
      "firstName": null,
      "id": "c4cb4582-1e97-4e3e-9d49-c744c8c1a229",
      "lastName": null,
      "state": "ACTIVE",
      "updatedAt": 2020-01-01T08:21:59.730Z,
    },
  ],
  "meta": Object {
    "count": 2,
    "totalCount": 6,
  },
}
`;

exports[`UsersQueries getUsers should return a paged list of users: #2 1`] = `
Object {
  "data": Array [
    User {
      "createdAt": 2020-01-01T08:21:59.730Z,
      "email": "user1@silvernext.com",
      "firstName": "Tom",
      "id": "49affd08-f461-4e80-899f-7612e971ff12",
      "lastName": "De Buyser",
      "state": "ACTIVE",
      "updatedAt": 2020-01-01T08:21:59.730Z,
    },
  ],
  "meta": Object {
    "count": 1,
    "totalCount": 1,
  },
}
`;

exports[`UsersQueries getUsers should return a paged list of users: #3 1`] = `
Object {
  "data": Array [
    User {
      "createdAt": 2019-01-01T08:32:59.730Z,
      "email": "user4@silvernext.com",
      "firstName": null,
      "id": "c356fce8-9f95-4c69-9e71-8f4d74f49b8f",
      "lastName": null,
      "state": "ACTIVE",
      "updatedAt": 2020-01-02T08:59:59.730Z,
    },
    User {
      "createdAt": 2020-01-01T08:21:59.730Z,
      "email": "development@icapps.com",
      "firstName": null,
      "id": "c4cb4582-1e97-4e3e-9d49-c744c8c1a229",
      "lastName": null,
      "state": "ACTIVE",
      "updatedAt": 2020-01-01T08:21:59.730Z,
    },
    User {
      "createdAt": 2020-01-01T08:21:59.730Z,
      "email": "user1@silvernext.com",
      "firstName": "Tom",
      "id": "49affd08-f461-4e80-899f-7612e971ff12",
      "lastName": "De Buyser",
      "state": "ACTIVE",
      "updatedAt": 2020-01-01T08:21:59.730Z,
    },
    User {
      "createdAt": 2020-01-01T08:21:59.730Z,
      "email": "user3@silvernext.com",
      "firstName": "Jens",
      "id": "db5e788c-2642-4c40-a63a-f13675048ddc",
      "lastName": "De Craecker",
      "state": "REGISTERING",
      "updatedAt": 2020-01-01T08:21:59.730Z,
    },
    User {
      "createdAt": 2020-01-01T10:21:59.730Z,
      "email": "user2@silvernext.com",
      "firstName": "Benno",
      "id": "45c2c554-b06a-4209-aba8-b097cb5f8566",
      "lastName": "Meysmans",
      "state": "ACTIVE",
      "updatedAt": 2020-01-05T08:21:59.730Z,
    },
    User {
      "createdAt": 2020-01-02T09:21:59.730Z,
      "email": "user5@silvernext.com",
      "firstName": null,
      "id": "af965f04-8c7d-4e20-b230-932b0ff3a33f",
      "lastName": null,
      "state": "REGISTERING",
      "updatedAt": 2020-01-02T09:21:59.730Z,
    },
  ],
  "meta": Object {
    "count": 6,
    "totalCount": 6,
  },
}
`;

exports[`UsersQueries getUsers should return an empty list if no users found 1`] = `
Object {
  "data": Array [],
  "meta": Object {
    "count": 0,
    "totalCount": 0,
  },
}
`;
